### ISSUE TITLE
Redundant entries in the DeclContexts

### ISSUE THREAD

**Issue by Xazax-hun** [Ericsson/clang#239]
Unfortunately, the ASTImporter sometimes adds redundant entries to some DeclContext.

For example a ClassTemplateDecl has a CXXRecordDecl as it's child.
The ClassTemplateDecl is added to the enclosing DeclContext in the original one but CXXRecordDecl is not.

The ASTImporter always adds CXXRecordDecl (and others like TypeAliasDecl) to the enclosing DeclContext in the "to" context, even if it is redundant, because this node is the child of a template node.

**Comment by Xazax-hun** [comment]
For example the original DeclContext might look like:
```
template <typename ...Ts> using X = int [sizeof...(Ts)];
template <typename ...Us> struct Y {
    X<Us..., int, double, int, Us...> f;
};
Y<float, int> declToImport;

StoredDeclsMap TranslationUnit 0x63b5c0
|-DeclarationName '__builtin_va_list'
| `-Typedef 0x671780 '__builtin_va_list'
|-DeclarationName '__builtin_ms_va_list'
| `-Typedef 0x63bfa0 '__builtin_ms_va_list'
|-DeclarationName '__NSConstantString'
| `-Typedef 0x63bf08 '__NSConstantString'
|-DeclarationName '__uint128_t'
| `-Typedef 0x63bbc0 '__uint128_t'
|-DeclarationName '__int128_t'
| `-Typedef 0x63bb50 '__int128_t'
|-DeclarationName 'Y'
| `-ClassTemplate 0x671af8 'Y'
|-DeclarationName 'X'
| `-TypeAliasTemplate 0x671988 'X'
`-DeclarationName 'declToImport'
  `-Var 0x69b678 'declToImport' 'Y<float, int>':'struct Y<float, int>'
```

The imported looks like:

```
using X = int [sizeof...(Ts)];
template <typename ...Ts> using X = int [sizeof...(Ts)];
struct Y {
};
template <typename ...Us> struct Y {
};
template<> struct Y<<float, int>>;

StoredDeclsMap TranslationUnit 0x6eca00
|-DeclarationName '__builtin_ms_va_list'
| `-Typedef 0x6ed3e0 '__builtin_ms_va_list'
|-DeclarationName '__builtin_va_list'
| `-Typedef 0x722430 '__builtin_va_list'
|-DeclarationName '__uint128_t'
| `-Typedef 0x6ed000 '__uint128_t'
|-DeclarationName '__int128_t'
| `-Typedef 0x6ecf90 '__int128_t'
|-DeclarationName '__NSConstantString'
| `-Typedef 0x6ed348 '__NSConstantString'
|-DeclarationName 'X'
| |-TypeAlias 0x722858 'X'
| `-TypeAliasTemplate 0x7228b8 'X'
`-DeclarationName 'Y'
  |-ClassTemplate 0x7226f0 'Y'
  `-CXXRecord 0x722488 'Y'
```

**Comment by Xazax-hun** [comment]
Proposed solution in: https://reviews.llvm.org/D42301

**Comment by martong** [comment]
Fixed by #255, #258, #270
